### PR TITLE
Migrating prombench trigger from GitHub Actions to Google Cloudbuild.

### PR DESCRIPTION
**Pull request by saketjajoo**
This migration is done to use GCP's cloudbuild service instead of GitHub Actions because of security implications, internal organization policy and the ease of implementation (to a certain extent). Using cloudbuild eliminates the need to store sensitive credentials in GitHub's environment.

### CODE REVIEW

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: can we use default substitution of $PROJECT_ID https://cloud.google.com/build/docs/configuring-builds/substitute-variable-values#using_default_substitutions

reduces the need to add it as a trigger substitution variable.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: not used

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: not used

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: why are we skipping this?

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: Do we want to expose our file name here? 

also I see versions/latest in the guide. Should we use latest in stead of 1?

ttps://cloud.google.com/build/docs/securing-builds/use-secrets

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: Need to have a graceful exit. Maybe a global variable called exit that you can check between steps.

https://cloud.google.com/build/docs/configuring-builds/pass-data-between-steps

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: _PR_NUMBER and _GIT_REPO are available in default subs https://cloud.google.com/build/docs/configuring-builds/substitute-variable-values#using_default_substitutions

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: nit: combine line 69-89 into one If using OR. return log `Requirements not met to trigger prom bench because ...`

will make code simpler to read

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: Just put the curl command here since it's called once. It will use less lines 

```
        curl --silent -L -X POST \
          -H "Accept: application/vnd.github+json" \
          -H "Authorization: Bearer $${GH_PAT}" \
          -H "X-GitHub-Api-Version: 2022-11-28" \
          "$${URL}" \
          -d '{"body": "'"$${BODY}"'"}'
```

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: Let's simplify this code a bit. Let's avoid nested ifs and repeated code like line 128 and 136. This will improve readability. 

How about this pseudo code. 
If there is a tag check if it exists
if no tag or tag doesn't exist  fetch latest 
else: use passed tag

```
	GCLOUD_TAG = ''
	if [[ "$${COLLECTOR_TAG_OR_CANCEL}" != '' ]]
	then
		GCLOUD_TAG=$(gcloud container images list-tags --filter="tags:$${COLLECTOR_TAG_OR_CANCEL}" --format=json $${PROMETHEUS_IMAGE_REPOSITORY})
	fi
	if [[ "$${GCLOUD_TAG}" == "[]" ]] || [[ "$${COLLECTOR_TAG_OR_CANCEL}" == '' ]]
		echo "Since the collector version is missing, the latest one would be used."
		DEFAULT_COLLECTOR_TAG=$(gcloud container images list-tags "$${PROMETHEUS_IMAGE_REPOSITORY}" --sort-by=UPDATE_TIME,TAGS | tail -1 | awk '{ print $2 }')
		echo "$${DEFAULT_COLLECTOR_TAG}" > /workspace/PROMETHEUS_IMAGE_VERSION
		export PROMETHEUS_IMAGE_VERSION="$${DEFAULT_COLLECTOR_TAG}"
	else
              echo "$${COLLECTOR_TAG_OR_CANCEL}" > /workspace/PROMETHEUS_IMAGE_VERSION
              export PROMETHEUS_IMAGE_VERSION="$${COLLECTOR_TAG_OR_CANCEL}"
         fi 
```

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: can you inline echo "false" > /workspace/CREATE_CLUSTER since `CREATE_CLUSTER` is not used anywhere else, no need for a variable.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: same comment as 158

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml:154]: can we move this before any gcloud command is run? this may cause issues for other projects in the future.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: can you break this into steps? I think it will improve readability. check go/tott/674 as a ref.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: can inline each export "$${ENV_VAR_NAME}=$(cat /workspace/$${ENV_VAR_NAME})"

can create each file `Setting environment variables` step to avoid the check on line 204

**Review by maxamins — COMMENTED** [review]
Added some comments to improve readability and reliability. Let me know your thoughts are.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: This is used in the [comment-monitor](https://github.com/GoogleCloudPlatform/prometheus-test-infra/blob/master/tools/commentMonitor/client.go#L105). Although we don't use comment monitor, the deployment may fail if the comment-monitor fails to start up. Hence, thought to keep it, but have it empty.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: Having the `ALLOW_LOOSE` parameter means the build does not return an error if there's a missing substitution variable or a missing substitution.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: Since the current set-up is not based on the [GitHub repo](https://cloud.google.com/build/docs/automating-builds/github/connect-repo-github?generation=2nd-gen), the variables get substituted at runtime. Once this PR gets merged, we can change these variables to use the default available substitutions (maybe as a next iteration).

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: Currently, the secrets in the secrets manager are versioned. Maybe we can use [version aliases](https://cloud.google.com/secret-manager/docs/assign-alias-to-secret-version) so that when secrets are rotated, the alias can point to the newer version and the "latest" version need not be updated in code.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: Having multiple checks instead of combining them into a single `if` gives us granularity while logging. We can check the logs to understand why something might have gone wrong if something goes wrong.
Combining everything in a single `if` means that even if the error message gets logged, it still needs to be decoded to understand why something went wrong (which exact requirement was not met).

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: > Maybe a global variable called exit that you can check between steps.

Maybe this might do the trick along with having [allowexitcodes](https://cloud.google.com/build/docs/build-config-file-schema#allowexitcodes). This way the steps can be skipped all the way to the end of the build.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: Although this is used only once in this step, a separate function was created to have it consistent with line #185. This way, if we need to change the `post_to_github` function, we need not read everything and just change the 2 same functions.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml:154]: The previous gcloud commands get the authentication credentials from GCE metadata server. However, we need to have credentials stored locally for prombench (GCP APIs called in prombench after this step). The reason that this is not before the first gcloud command is that we only need this file if all the previous steps succeed. Otherwise, we need not store the credentials locally if any step fails (which fails the entire build).

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: Yes, I agree. But breaking it into steps means that the environment variables that got changed in previous steps are reset and have the default (stale) values. If so, we'll need to have the `get_or_set_env_var` function in each step to get the updated value.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml]: > can create each file Setting environment variables step to avoid the check on line 204

This will make the whole thing consistent, which is good! But, then if later on the environment variables are added or deleted, the file creation step/function would need to be updated as well. Contributors might forget to update the file creation step/function as there are only 3 variables that get updated (as of now). Rest all the env vars are used as is.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: maybe add a comment explaining reasoning behind flags.

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [inline .cloudbuild/prombench.yaml:2]: Will move this section after steps.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [reply .cloudbuild/prombench.yaml]: ignore _PR_NUMBER only available for PR triggers

Comment (maxamins) [reply .cloudbuild/prombench.yaml]: also ignor github repo. Only project_id can be subbed

File: .cloudbuild/prombench.yaml
Comment (maxamins) [reply .cloudbuild/prombench.yaml]: OK, makes sense. Also names are generic enough.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: Do we need this check for every global variable?

 What if we have a step to create files with default values. That way we know the file exists for sure.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml:117]: 101 and 104 can be on line 107 as echo "$${PROMETHEUS_IMAGE_VERSION}" > /workspace/PROMETHEUS_IMAGE_VERSION

**Review by maxamins — COMMENTED** [review]
Looking very good! reviewed half, will reviewed half later.

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: do you actually need export?

File: .cloudbuild/prombench.yaml
Comment (maxamins) [inline .cloudbuild/prombench.yaml]: add comment explaining it

**Review by maxamins — COMMENTED** [review]
Thanks for making the many changes 👍 ! finished reviewing!

File: .cloudbuild/prombench.yaml
Comment (bwplotka) [inline .cloudbuild/prombench.yaml:42]: Is `_` prefix is some pattern? It's kind of unreadable and odd.  If there's reason, do you mind adding comment why?

File: .cloudbuild/prombench.yaml
Comment (bwplotka) [inline .cloudbuild/prombench.yaml:48]: What's PAT?

**Review by bwplotka — APPROVED** [review]
Thanks! Added a few readability comments/questions but overall LGTM.

Looks more complex and we again put something unique to our fork which will be pain to maintainer, but if there's no other way 🤷🏽

File: .cloudbuild/prombench.yaml
Comment (TheSpiritXIII) [reply .cloudbuild/prombench.yaml:48]: Personal Access Token?

Comment (saketjajoo) [reply .cloudbuild/prombench.yaml:48]: It's GitHub bot's Personal Access Token (to authenticate to GitHub).

File: .cloudbuild/prombench.yaml
Comment (saketjajoo) [reply .cloudbuild/prombench.yaml:42]: One of the [rules for declaring user-defined substitution variables](https://cloud.google.com/build/docs/configuring-builds/substitute-variable-values#using_user-defined_substitutions) is that they must start with `_`.

**Pull request merged**
